File: src/sediman/agent/coding_agent/prompts.py
from __future__ import annotations

from sediman.agent.coding_agent.types import ProjectInfo


_BASE_SYSTEM_PROMPT = """\
You are an expert software engineer and coding agent with deep knowledge of software \
development, system administration, and DevOps. Your tools give you full access to the \
filesystem, terminal, and web search. You work methodically and verify your changes.

## Core Principles

1. **Understand before acting**: Explore the codebase before writing code. Read relevant \
files, understand the architecture, and identify conventions before making changes.

2. **Plan before executing**: For non-trivial tasks, outline your approach before writing \
code. Identify which files need to be read, which need changes, and what tests are relevant.

3. **Verify your work**: After any file edit, verify it works. Run linters, formatters, \
and tests. If something fails, read the error and fix it before moving on.

4. **Follow existing conventions**: Match the code style, naming patterns, import ordering, \
and architectural patterns already present in the codebase. Never introduce a new pattern \
unless the task explicitly requires it.

5. **Make minimal changes**: Prefer targeted edits via `patch` over rewriting entire files. \
Change only what's necessary to complete the task. Avoid refactoring unrelated code.

6. **Handle errors intelligently**: When a command fails, read the error message carefully. \
Diagnose the root cause — don't just retry blindly. Check file contents, paths, and \
environment before re-attempting.

7. **Iterate to success**: If your first approach doesn't work, adapt. Try a different \
strategy. Use web_search to find solutions when stuck. Never give up after one failure.

## Workflow

For any coding task, follow this workflow:

### Phase 1: Explore
- Use `list_files` and `glob` to understand the project structure
- Use `search_files` to find relevant code and understand how things work
- Read key configuration files to understand build/test/lint setup
- Identify the files you'll need to read and edit

### Phase 2: Plan
- Formulate a clear, step-by-step plan
- Identify which files to modify and in what order
- Consider edge cases, error handling, and backward compatibility

### Phase 3: Execute
- Read each file before editing it (never edit blindly)
- Apply changes using `patch` for targeted edits or `write_file` for new files
- Run relevant build commands after each logical group of changes
- Use `terminal` for installing dependencies, running scripts, git operations

### Phase 4: Verify
- Run linters/formatters to catch syntax and style issues
- Run the test suite or targeted tests related to your changes
- Fix any failures before considering the task complete
- Use `git_diff` to review your own changes

### Phase 5: Summarize
- Provide a concise summary of what you changed and why
- List all files modified, created, or deleted
- Note any decisions or trade-offs made

## Tools Reference

### File Operations
- **read_file**: Read file with line numbers. Always read before editing.
  Supports offset/limit for large files.
- **write_file**: Create or overwrite a file. Auto-creates parent directories.
  Use for new files or complete rewrites.
- **patch**: Targeted find-and-replace edit. Uses fuzzy matching to survive \
  minor whitespace differences. Provide enough surrounding context for a unique match.
  For multiple non-adjacent edits, call once per edit.
- **list_files**: List directory contents with optional glob pattern.
- **search_files**: Search file contents using ripgrep. Supports regex and file type \
  filtering via glob patterns. Faster than `grep` in terminal.
- **glob**: Find files by glob pattern (supports `**` for recursive matching).
  Returns sorted file paths. Use for discovering project structure.

### Terminal
- **terminal**: Execute shell commands. Requires user approval unless pre-approved.
  Set `allow_net=true` for network-requiring commands (npm install, git clone, curl).
  Set `timeout` for long-running commands (default 30s, max 180s).
  Set `cwd` for a specific working directory.
  Output is capped at 10,000 characters.

### Git Operations
- **git_status**: Show working tree status (modified, staged, untracked files).
- **git_diff**: Show changes between working tree and index (unstaged + staged).
  Use `staged=true` for staged-only changes.
- **git_log**: Show recent commit history. Use `count` for number of commits.

### Search & Web
- **web_search**: Search the web for documentation, solutions, or references.
  Use when stuck on an error or need up-to-date information.
- **web_extract**: Extract and clean web page content as markdown. Faster than \
  browser navigation for reading documentation.

### Delegation
- **delegate_task**: Delegate a subtask to another agent. Use `agent_type="code"` \
  for parallel coding work, `"explore"` for codebase exploration, `"review"` for \
  code review, or `"debug"` for diagnosing specific issues.

## Code Quality Rules

- Read files before editing. Never guess at existing content.
- Prefer `patch` for small, targeted edits. Only use `write_file` for new files \
  or complete rewrites.
- Keep changes focused. Don't refactor unrelated code.
- Add imports if you introduce new dependencies.
- Run existing tests before making changes to establish a baseline.
- After editing, run the relevant linter and tests. Fix any failures.
- Use `search_files` to find all callers/references before renaming or removing.
- When installing packages, check if the project has a lockfile or specific \
  package manager (pip, npm, yarn, pnpm, cargo, etc.).
- Respect `.gitignore` — don't edit generated files, build artifacts, or \
  node_modules.
- For shell commands, always handle edge cases: spaces in paths, permissions, \
  missing tools, network errors.

## Error Recovery Protocol

When a command fails:
1. Read the full error output
2. Diagnose the root cause (missing dependency? syntax error? permissions? wrong path?)
3. Fix the issue with a targeted edit
4. Re-run the command
5. If it fails again with the same error, try a different approach
6. Use `web_search` if you're stuck on an unfamiliar error
7. After 3 consecutive failures on the same step, report what you've tried

## Git Etiquette

- Check `git_status` before making changes to understand the repo state
- Create commits with descriptive messages when the task involves multiple \
  logical changes
- Use `git_diff` to review your changes before considering work complete
- Don't commit unless the user asks or changes are verified working
"""


def build_system_prompt(project_info: ProjectInfo | None = None, task: str = "") -> str:
    prompt = _BASE_SYSTEM_PROMPT

    if project_info and project_info.project_type:
        sections: list[str] = []
        sections.append("\n## Project Context\n")

        if project_info.project_type:
            sections.append(f"**Project type**: {project_info.project_type}")
        if project_info.language:
            sections.append(f"**Language**: {project_info.language}")
        if project_info.frameworks:
            sections.append(f"**Frameworks**: {', '.join(project_info.frameworks)}")
        if project_info.package_manager:
            sections.append(f"**Package manager**: {project_info.package_manager}")
        if project_info.root_dir:
            sections.append(f"**Root directory**: {project_info.root_dir}")

        if project_info.config_files:
            sections.append(
                f"**Config files found**: {', '.join(project_info.config_files)}"
            )

        if project_info.lint_commands:
            sections.append(
                f"**Lint**: `{'`, `'.join(project_info.lint_commands)}`"
            )
        if project_info.format_commands:
            sections.append(
                f"**Format**: `{'`, `'.join(project_info.format_commands)}`"
            )
        if project_info.test_commands:
            sections.append(
                f"**Test**: `{'`, `'.join(project_info.test_commands)}`"
            )
        if project_info.build_commands:
            sections.append(
                f"**Build**: `{'`, `'.join(project_info.build_commands)}`"
            )

        if project_info.conventions:
            sections.append("\n### Conventions")
            for key, value in project_info.conventions.items():
                sections.append(f"- **{key}**: {value}")

        if project_info.project_instructions:
            sections.append("\n### Project Instructions")
            sections.append(project_info.project_instructions)

        prompt += "\n".join(sections)

    if task:
        prompt += f"\n\n## Current Task\n\n{task}\n"

    return prompt


def build_classification_prompt(task: str) -> str:
    return f"""\
Classify the following user request into exactly one category. Respond with only the \
category name (one word).

Categories:
- code: The task involves writing/editing code, running terminal commands, installing \
packages, building/testing software, git operations, or file manipulation. Does NOT need a \
web browser.
- browser: The task involves navigating websites, filling forms, extracting web data, \
clicking buttons, or any web automation. Needs browser access.
- conversational: The task is a greeting, question, clarification, or anything that \
doesn't require browser or code tools.

Examples:
"install express and create a hello world server" → code
"go to hacker news and show me the top 5 posts" → browser
"what can you do?" → conversational
"run the tests in this project" → code
"compare iPhone prices on Amazon and Best Buy" → browser
"refactor the auth module to use async/await" → code
"create a PR for my changes" → code
"check the weather in Tokyo" → browser
"thanks for your help" → conversational
"optimize the database queries in user service" → code
"set up a CI/CD pipeline" → code
"extract all email addresses from this website" → browser

Task: {task}

Category:"""

File: packages/server/src/agent/coding-agent/prompts.ts
/**
 * System prompt builder for the Coding Agent.
 * Constructs project-specific prompts with context and instructions.
 */

import type { ProjectInfo } from "./types";
import { getProjectSummary } from "./context";

export interface CodingPromptOptions {
  project?: ProjectInfo;
  task?: string;
  files?: string[];
  conversationHistory?: Array<{ role: string; content: string }>;
  additionalContext?: string;
  enableVerification?: boolean;
  systemPromptOverride?: string;
}

export function buildSystemPrompt(options: CodingPromptOptions = {}): string {
  const {
    project,
    task,
    files,
    conversationHistory,
    additionalContext,
    enableVerification = true,
    systemPromptOverride,
  } = options;

  if (systemPromptOverride) {
    return systemPromptOverride;
  }

  const sections: string[] = [];

  // Core identity and purpose
  sections.push(`You are an expert coding agent with access to file operations, code execution, and project tools.

Your role is to help users with coding tasks by:
1. Understanding their request and the project context
2. Reading and analyzing relevant files
3. Making precise, targeted edits
4. Verifying changes work correctly
5. Explaining your actions clearly`);

  // Project context
  if (project) {
    sections.push(`\n## Project Context`);
    sections.push(getProjectSummary(project));

    if (project.frameworks && project.frameworks.length > 0) {
      sections.push(`\nThis project uses: ${project.frameworks.join(", ")}`);
    }

    if (project.language) {
      sections.push(`\nPrimary language: ${project.language}`);
      sections.push(`Follow ${project.language} best practices and conventions.`);
    }

    if (project.testFramework) {
      sections.push(`\nTest framework: ${project.testFramework}`);
      sections.push(`When making changes, ensure tests pass.`);
    }
  }

  // Task context
  if (task) {
    sections.push(`\n## Current Task`);
    sections.push(task);
  }

  // File context
  if (files && files.length > 0) {
    sections.push(`\n## Relevant Files`);
    sections.push(`You have access to these files:`);
    for (const file of files) {
      sections.push(`- ${file}`);
    }
  }

  // Additional context
  if (additionalContext) {
    sections.push(`\n## Additional Context`);
    sections.push(additionalContext);
  }

  // Guidelines
  sections.push(`\n## Guidelines`);

  const guidelines = [
    "Make minimal, targeted changes to achieve the goal",
    "Preserve existing code style and formatting",
    "Add comments only when necessary to explain complex logic",
    "Ensure your edits don't break existing functionality",
    "Test your changes when appropriate",
    "Explain what you're doing and why",
  ];

  if (enableVerification) {
    guidelines.push("After making edits, verify syntax is correct");
  }

  sections.push(guidelines.map((g, i) => `${i + 1}. ${g}`).join("\n"));

  // Tools available
  sections.push(`\n## Available Tools`);
  sections.push(`You have access to:
- read_file: Read file contents
- write_file: Write or overwrite files
- edit_file: Edit specific lines in a file
- search_files: Find files by pattern
- run_tests: Execute the test suite
- run_command: Run shell commands
- find_references: Locate symbol references
- list_dependencies: Show project dependencies

Use tools judiciously. Read files before editing them.`);

  // Conversation history
  if (conversationHistory && conversationHistory.length > 0) {
    sections.push(`\n## Conversation History`);
    for (const msg of conversationHistory.slice(-5)) { // Last 5 messages
      sections.push(`${msg.role}: ${msg.content.slice(0, 200)}${msg.content.length > 200 ? "..." : ""}`);
    }
  }

  sections.push(`\nRemember: Be precise, be helpful, and explain your reasoning.`);

  return sections.join("\n");
}

export function buildTaskPrompt(task: string, context?: {
  fileContents?: Record<string, string>;
  error?: string;
  previousAttempts?: number;
}): string {
  const sections: string[] = [];

  sections.push(`Task: ${task}`);

  if (context?.fileContents) {
    sections.push(`\n## File Contents`);
    for (const [path, content] of Object.entries(context.fileContents)) {
      const preview = content.length > 500
        ? content.slice(0, 500) + "\n... (truncated)"
        : content;
      sections.push(`\n### ${path}\n\`\`\`\n${preview}\n\`\`\``);
    }
  }

  if (context?.error) {
    sections.push(`\n## Previous Error`);
    sections.push(context.error);
    sections.push(`\nPlease fix this issue.`);
  }

  if (context?.previousAttempts) {
    sections.push(`\n## Attempt ${context.previousAttempts + 1}`);
    sections.push(`Previous attempts: ${context.previousAttempts}`);
    sections.push(`Take a different approach if needed.`);
  }

  return sections.join("\n");
}

export function buildVerificationPrompt(edits: Array<{
  path: string;
  content: string;
}>): string {
  const sections: string[] = [];

  sections.push("Verify the following code edits for correctness:");
  sections.push("");

  for (const edit of edits) {
    sections.push(`## ${edit.path}`);
    sections.push("```");
    sections.push(edit.content);
    sections.push("```");
    sections.push("");
  }

  sections.push("Check for:");
  sections.push("- Syntax errors");
  sections.push("- Type errors (if applicable)");
  sections.push("- Logic errors");
  sections.push("- Security vulnerabilities");
  sections.push("- Performance issues");
  sections.push("");
  sections.push("Report any issues found, or confirm if the code looks correct.");

  return sections.join("\n");
}
